File: server/lib/item_parser.js
class ItemParser {
  constructor(itemString) {
    this.item      = {};
    this.fragments = itemString.split('--------');

    this._parseFragments();
  }

  // @private

  _parseFragments() {
    this.fragments.forEach((fragment) => {
      let rows = fragment.split("\n").
                          filter((row) => row.length);

      this._parseSimpleAttributes(rows);
      this._parseSimpleStrings(rows);
    });
  }

  _parseSimpleStrings(rows) {
    let plainTextAttributes = ['Shaper', 'Elder', 'Corrupted'];

    rows.forEach((row) => {
      // Check to see if our row is one of the plain text attributes in our list, if it
      // is not then we can skip it
      if (!plainTextAttributes.includes(row)) { return false; }

      // Add the key and value we found to our item
      this.item[row] = true
    });
  }

  _parseSimpleAttributes(rows) {
    let regexp = RegExp(/^[a-zA-Z1-9 ]*: .*$/);

    rows.forEach((row) => {
      // Check if this is an easy attribute to find, if not return
      // false so we can check the next line
      if (!regexp.test(row)) { return false; }

      let rowData    = row.split(': ');

      // Add the key and value we found to our item
      this.item[rowData[0]] = rowData[1]
    });
  }
}

module.exports = ItemParser;

File: server/lib/item_parser.rb
class ItemParser
  SEPERATOR = '--------'

  def initialize(item_string)
    @item      = { modifiers: [] }
    @fragments = item_string.split(SEPERATOR)

    parse_fragments
  end

  private

    def parse_fragments()
      @fragments.each do |fragment|
        rows = fragment.split("\n").compact

        parse_simple_attributes(rows)
        parse_simple_strings(rows)
        parse_modifiers(rows)
      end
    end

    def parse_simple_attributes(rows)
      regexp = /^[a-zA-Z1-9 ]*: .*$/

      rows.each do |row|
        next unless regexp =~ row

        key, value = row.split(': ')

        @item[key] = value
      end
    end

    def parse_simple_strings(rows)
      plain_text_attributes = %w[Shaper Elder Corrupted]

      rows.each do |row|
        next unless plain_text_attributes.include?(row)

        @item[row] = true
      end
    end

    def parse_modifiers(rows)
      regexp = /[-+]?[0-9]*\.?[0-9]{1,2}/

      rows.each do |row|
        row = row.gsub(regexp, '{_}').
                  gsub('%', '[%]').
                  gsub('\'', '\'\'')

        modifier_query = "SELECT TOP (1) ModifierID FROM ModifierString WHERE String LIKE #{row}'";

        puts Database.new
      end
    end

end